src/librustc_mir/interpret/place.rs

```diff
@@ -1040,21 +1040,18 @@ where
     ) -> InterpResult<'tcx> {
         let variant_scalar = Scalar::from_u32(variant_index.as_u32()).into();
 
+        if dest.layout.for_variant(self, variant_index).abi.is_uninhabited() {
+            throw_ub!(InvalidDiscriminant(variant_scalar));
```

Comment:
So I think this is a misuse of `InvalidDiscriminant`, because there's no discriminant being *read* (which is why I assume it takes a `Scalar`).

I would prefer a different error for "tried to create uninhabited variant" (which could contain the type and variant index, let's say), which should generally be impossible outside of dead code.

And keep `InvalidDiscriminant` for something that can totally happen with unsafe code.

Reply:
> which should generally be impossible outside of dead code.

Should it be `InvalidProgram` then in terms of error category, instead of `UndefinedBehavior`?

Reply:
Or maybe UB actually makes sense (rustc just happens to not allow the user to write such MIR). But the issue is that layout code doesn't actually provide a layout for uninhabited variants, so there's no reasonable way to implement `SetDiscriminant`. So maybe `throw_ub!(Unreachable)`? I assume codegen also emits an "unreachable" for LLVM here.

Reply:
> I assume codegen also emits an "unreachable" for LLVM here.

No, it does nothing, `unreachable` might make more sense though.

Actually, I'm not sure how right I am about the dead code part, since the MIR creating that uninhabited variant is still left around.

Presumably it requires a call that returns an uninhabited type, but not `!` (which typeck would know about).

So it's "dynamically dead" code, in that the function could never return, but this is not encoded in the CFG.

Reply:
> No, it does nothing, unreachable might make more sense though.

Yeah I'd be in favor of keeping Miri and codegen in sync here.

Reply:
I will change this to `throw_ub!(Unreachable)` then.